scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -1667,3 +1670,459 @@ cdef FibonacciNode* remove_min(FibonacciHeap* heap) noexcept:
+            # Add the found path to arrays of candidates
+            if (
+                total_distance != INFINITY
+                and _yen_is_path_in_candidates(candidate_predecessors,
+                                               shortest_paths_predecessors[k-1], 
+                                               predecessor_matrix[0],
+                                               spur_node, sink) == 0
+            ):
+                # Find the index to insert the new path
+                short_path_idx = tmp_i = NULL_IDX
+                tmp_d = -INFINITY # maximal distance in potential distances array
+                for i in range(candidate_distances.shape[0]):
+                    if candidate_distances[i] == INFINITY:
+                        short_path_idx = i
+                        break
+                    elif candidate_distances[i] > tmp_d:
+                        tmp_d = candidate_distances[i]
+                        tmp_i = i
+                if short_path_idx ==  NULL_IDX and total_distance < tmp_d:
+                    short_path_idx = tmp_i
+
+                if short_path_idx != NULL_IDX:
+                    candidate_distances[short_path_idx] = total_distance
+                    # Reset candidate_predecessors[short_path_idx]
+                    candidate_predecessors[short_path_idx, :] = NULL_IDX
+                    # Fill original path
+                    node = spur_node
+                    while node != NULL_IDX:
+                        candidate_predecessors[short_path_idx, node] = (
+                            shortest_paths_predecessors[k-1, node]
+                        )
+                        node = shortest_paths_predecessors[k-1, node]
+
+                    # Fill spur path
+                    node = sink
+                    while node != spur_node:
+                        candidate_predecessors[short_path_idx, node] = (
+                            predecessor_matrix[0, node]
+                        )
+                        node = predecessor_matrix[0, node]
+
+           # ---------------------------------------------------
+            # Restore graph weights
+            node = spur_node
+            while node != NULL_IDX:
+                csr_weights[csr_indptr[node]: csr_indptr[node + 1]] = (
+                    original_weights[csr_indptr[node]: csr_indptr[node + 1]]
+                )
+                if not directed:
+                    csrT_weights[csrT_indptr[node]: csrT_indptr[node + 1]] = (
+                        originalT_weights[csrT_indptr[node]: csrT_indptr[node + 1]]
+                    )
+                node = shortest_paths_predecessors[k-1, node]
+
+
+        # ---------------------------------------------------
+        # Find shortest path in candidates and add to result arrays
+        tmp_d = INFINITY # Minimal distance in potential distances array
+        short_path_idx = NULL_IDX
+        for i in range(candidate_distances.shape[0]):
```

Comment:
This has already been addressed in a previous [comment](https://github.com/scipy/scipy/pull/20187#issuecomment-1987862567). The conclusion is that, yes, it would increase performance, and should be done, however, this PR is already quite large, and our heap data structure is designed specifically for Dijkstra's algorithm so it would require a significant rework. 

My _opinion_ on this is that we should do this in a following PR

Reply:
Plus it would probably be better to have #17019 merged first before doing that

Reply:
I agree with @j-bowhay .
A major consideration when choosing not to implement heaps in this PR was that the module is compiled to C and that caused further restrictions.
As https://github.com/scipy/scipy/pull/17019 moves the module compilation to cpp, and removes the local heap structure. I believe it would be more straightforward to convert `candidate_distances` to a heap once the features of cpp are available (e.g., priority queue).